### PR TITLE
Dev Console

### PR DESCRIPTION
**Pull request by FroMage**
We can merge/squash some commits.

This has been rebased on master, and is missing documentation I'll write ASAP.

### CODE REVIEW

**Comment by gsmet** [comment]
@FroMage thanks, glad to see this taking form! And yes, this will need a proper squash.

Also how does it look like? I was wondering if maybe I should have a look and make use of Semantic UI to get something nice until we get help from the design team? Or maybe you already took care of that? The only screenshot I saw was from a very long time ago.

**Comment by FroMage** [comment]
Squashed properly somewhat, though we lost all the DEV Console framework history, as usual.

Here's a screenshot: 
Screenshot from 2020-12-15 17-54-25

It looks pretty much like the current codestart landing page, so IMO that's ready for merging. I hope we can get fancy visuals later, but it should not delay extension authors, though we likely won't want to blog about this before we get great styling.

You can add semantic UI if you want, as long as it doesn't make this more complex than the current Bootstrap :)

**Comment by gsmet** [comment]
If you're already using Bootstrap, it's probably better to keep it I suppose as I think that's what the design team uses.

Semantic UI has this nice notion of cards: https://semantic-ui.com/views/card.html (there are tons of examples), which would be quite practical. I suppose we could find a Boostrap equivalent.

**Comment by FroMage** [comment]
Semantic UI looks fine, but I can't say whether it'd be more powerful than the Bootstrap cards https://getbootstrap.com/docs/5.0/components/card/

I did just notice that Bootstrap 5 is out now, so I should probably upgrade before merging :(

**Comment by gsmet** [comment]
I wouldn't upgrade before merging. Improving the UI and the bikeshedding can be done later!

**Comment by gsmet** [comment]
So on Semantic UI vs Bootstrap, Semantic UI is usually more polished design-wise. The newest version of Boostrap are quite rough to be as much customizable as possible.

But Semantic UI is not as well maintained as Bootstrap so I suppose Bootstrap is a safer choice.

**Comment by FroMage** [comment]
Docs have been added, and a PR against the website is at https://github.com/quarkusio/quarkusio.github.io/pull/815

**Comment by FroMage** [comment]
CI tells me this:

```
Caused by: org.apache.maven.plugin.MojoFailureException: File '/home/runner/work/quarkus/quarkus/core/runtime/src/main/java/io/quarkus/runtime/TemplateHtmlBuilder.java' has not been previously formatted.  Please format file and commit before running validation!
```

Oh damn, I keep building quickly, that's why it's not formatted, DAMNIT

**Comment by FroMage** [comment]
OK I'll fix this later today, damnit

**Comment by FroMage** [comment]
CI failed with this:

```
2020-12-16T15:03:24.2540332Z [INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-no-runtime-deps) @ quarkus-container-image-deployment ---
2020-12-16T15:03:24.2561564Z [WARNING] Rule 0: io.quarkus.enforcer.BansRuntimeDependency failed with message:
2020-12-16T15:03:24.2570878Z 1 illegal runtime dependencies found that have to be replaced with their -deployment counterparts:
2020-12-16T15:03:24.2574398Z 	quarkus-vertx-http
```

I've never understood these enforcer rules. Do you have any idea @stuartwdouglas @geoand ?

**Comment by geoand** [comment]
Oh darn...

Basically with this we now need runtime dependencies in the `deployment` module in order to support the dev console... 
But the enforcer prohibits this... 

@gastaldi @famod what should we do in this case?

**Comment by gastaldi** [comment]
@geoand why not depend on the `vertx-http-deployment` module instead?

**Comment by geoand** [comment]
Because we don't want to the bring in the entite extension (it doesn't make sense for example for the container-image extension to depend on the vert.x extension) - we want vert.x (optionally) so we can simply define routes for the dev console.

**Comment by famod** [comment]
@geoand Are all these runtime dependencies `<optional>true</optional>`?
If yes, we could theoretically adjust the rule to ignore such optional dependencies.

@gsmet WDYT, would this ^ an acceptable "softening" of the genreal mantra?

/cc @aloubyansky

**Comment by geoand** [comment]
@famod exactly

**Comment by famod** [comment]
A more strict approach would be to opt into this softening on a per module basis, ideally with a property.
While a rule can have config parameters, the downside of that would be that you basically have to redefine the entire rule in each module. A property would be more "lightweight" (but maybe less self-explaining).

**Comment by geoand** [comment]
I think that softening the rule to allow optional dependencies would be fine for now.

But I don't really have a strong opinion

**Comment by famod** [comment]
Btw, what about the build order? Are there (or will there be) any tests in e.g. `extensions/cache/runtime` that during "test runtime" need the respective deployment artifact (`vertx-web-deployment` and `vertx-http-deployment` for this specific example)?

**Comment by geoand** [comment]
Good question.

I don't think we have anything now, but I assume we will add some in the future

**Comment by famod** [comment]
Then we'd need those dreaded "minimal deployment dependencies" in the respective extension runtime modules.
Otherwise incremental/partial builds (manually via `-pl` or automatically via GIB) would operate on an inconsistent dependency graph. Same for parallelized test execution, in case that's ever going to be introduced.

**Comment by geoand** [comment]
Good point! Can you add an example of that so @FroMage can properly update the PR?

**Comment by famod** [comment]
Sure, but TBH I'd like to avoid doing so. I'm wondering whether such tests could be limited to `integration-tests/...`?
All IT modules do already have to define those minimal deployment dependencies, which is enforced by `RequiresMinimalDeploymentDependency` in `quarkus-integration-tests-parent`.

**Comment by geoand** [comment]
I think we can definitely do that - we just need to keep it in mind 😎

**Comment by famod** [comment]
I guess only time will tell whether this approach works. Not sure this could be validated automatically in some way... (e.g.: "don't pull deployment artifacts when running tests in a runtime module").

Btw, you'll have to redefine those optional dependencies in the respective IT module since they are not propagated.

**Comment by geoand** [comment]
> Btw, you'll have to redefine those optional dependencies in the respective IT module since they are not propagated.

👌👍

**Comment by famod** [comment]
**I'm very sorry but I think I mixed up a few things here!**
This is really about _none `optional`_ runtime dependencies in deployment modules:
https://github.com/quarkusio/quarkus/pull/13899/files?file-filters%5B%5D=.xml#diff-caddfcc0ef3f8661e3baacf6eafa48f6ef406c895c087b8e1fe10fab852867ee

So ignoring optional dependencies in the rule does not help at all. 😞

**Comment by aloubyansky** [comment]
Do we really need the vertx runtime extension artifact as a dep in every deployment module now?
If it's only for the dev mode, we could enhance the app model resolver to inject it in that mode only.
However, adding the runtime artifact isn't looking good. How about extracting the necessary classes from the vertx runtime artifact to a common jar instead?

**Comment by geoand** [comment]
We did it this way because it was the easiest thing to get going.

Enhancing the resolved for dev mode sounds fine to me.
Extracting the necessary classes into a common jar doesn't sound feasible, as those classes are part of vert.x itself.

In short, it would be great to have a proper solution, but it would also be nice to have the dev console in at some point because there is plenty of interest in it and it keeps get pushed back in time

**Comment by aloubyansky** [comment]
I meant extracting the necessary classes from the vertx runtime artifact to a separate module which would depend on the necessary vertx artifacts. Wouldn't that work?

**Comment by aloubyansky** [comment]
I see there already is quarkus-vertx-http-dev-console-spi. Why is the runtime artifact still needed? Can the SPI one be made to provide enough?

**Comment by geoand** [comment]
Ah yeah, that would likely work (although I haven't tried it).
Perhaps Stef and give it a shot tomorrow

**Comment by stuartwdouglas** [comment]
I have added a SPI module, lets see what CI says.

**Comment by FroMage** [comment]
CI says:

```
2020-12-17T03:04:02.9286896Z [INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.191 s - in io.quarkus.vertx.http.cors.CORSDisabledHandlerTestCase
2020-12-17T03:04:03.0629214Z [INFO] 
2020-12-17T03:04:03.0632467Z [INFO] Results:
2020-12-17T03:04:03.0669991Z [INFO] 
2020-12-17T03:04:03.0670454Z [ERROR] Failures: 
2020-12-17T03:04:03.0671534Z [ERROR]   HotReloadWithFilterTest.testAddFilter:58 1 expectation failed.
2020-12-17T03:04:03.0672597Z Expected status code <200> but was <500>.
2020-12-17T03:04:03.0672947Z 
2020-12-17T03:04:03.0673951Z [ERROR]   HotReloadWithFilterTest.testFilterChange:27 1 expectation failed.
2020-12-17T03:04:03.0675083Z Expected status code <200> but was <500>.
2020-12-17T03:04:03.0675428Z 
2020-12-17T03:04:03.0676368Z [ERROR]   HotReloadWithRouteTest.testRouteChange:23 1 expectation failed.
2020-12-17T03:04:03.0677450Z Expected status code <200> but was <500>.
2020-12-17T03:04:03.0677792Z 
2020-12-17T03:04:03.0678154Z [INFO] 
2020-12-17T03:04:03.0678661Z [ERROR] Tests run: 104, Failures: 3, Errors: 0, Skipped: 1
```

**Comment by stuartwdouglas** [comment]
That test is using  the path /dev, and the 'legacy redirect' stuff is now redirecting /dev.

As this is not a legacy thing I have added the ability to disable those redirects for new endpoints.

File: extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/devconsole/CacheDevConsoleRecorder.java
Comment (gwenneg) [inline extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/devconsole/CacheDevConsoleRecorder.java:27]: If this method is the only reason why `CacheManagerImpl.allCaches` was added, then I'd suggest to iterate over the caches names using `CacheManager.getCacheNames` and then retrieve the desired cache using `CacheManager.getCache`.

Comment (FroMage) [reply extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/devconsole/CacheDevConsoleRecorder.java:27]: Probably, yes.

**Comment by FroMage** [comment]
CI fails on Windows:

```
2020-12-18T02:43:30.0756523Z 	[error]: Build step io.quarkus.vertx.http.deployment.devmode.console.DevConsoleProcessor#buildTimeTemplates threw an exception: java.nio.file.InvalidPathException: Illegal char <:> at index 2: /C:/Users/runneradmin/.m2/repository/io/quarkus/quarkus-vertx-http-deployment/999-SNAPSHOT/quarkus-vertx-http-deployment-999-SNAPSHOT.jar
2020-12-18T02:43:30.0760811Z 	at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
2020-12-18T02:43:30.0763297Z 	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
2020-12-18T02:43:30.0764952Z 	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
2020-12-18T02:43:30.0766805Z 	at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
2020-12-18T02:43:30.0768572Z 	at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:229)
2020-12-18T02:43:30.0769752Z 	at java.base/java.nio.file.Path.of(Path.java:147)
2020-12-18T02:43:30.0770559Z 	at java.base/java.nio.file.Paths.get(Paths.java:69)
2020-12-18T02:43:30.0773074Z 	at io.quarkus.vertx.http.deployment.devmode.console.DevConsoleProcessor.collectTemplates(DevConsoleProcessor.java:338)
2020-12-18T02:43:30.0777761Z 	at io.quarkus.vertx.http.deployment.devmode.console.DevConsoleProcessor.buildTimeTemplates(DevConsoleProcessor.java:180)
```

File: extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/devconsole/CacheDevConsoleRecorder.java
Comment (geoand) [reply extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/devconsole/CacheDevConsoleRecorder.java:27]: Yeah

Comment (FroMage) [reply extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/devconsole/CacheDevConsoleRecorder.java:29]: Rewrote this, is it better?

**Comment by FroMage** [comment]
Rebased on master, rewrote cache impl, added a tentative windows fix.

File: extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CaffeineCacheSupplier.java
Comment (gwenneg) [inline extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CaffeineCacheSupplier.java:19]: You don't need to inject the impl anymore:
```suggestion
        CacheManager cacheManager = cacheManager();
```

File: extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/AbstractCache.java
Comment (gwenneg) [inline extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/AbstractCache.java:14]: This is really necessary (the names collection could be sorted before retrieving the cache instances), but it doesn't hurt to have this method public so I'm fine with either.

File: extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/devconsole/CacheDevConsoleRecorder.java
Comment (gwenneg) [reply extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/devconsole/CacheDevConsoleRecorder.java:29]: Yes, thanks!

File: extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/AbstractCache.java
Comment (FroMage) [reply extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/AbstractCache.java:14]: It's used in the view.

Comment (gwenneg) [reply extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/AbstractCache.java:14]: Oh sorry, I missed that. Then it's perfectly fine :)

**Comment by aloubyansky** [comment]
So the deployment artifacts should be bringing in `quarkus-vertx-http-dev-console-spi` and the runtime ones `vertx-web` and `quarkus-vertx-http`?
IMO, it'd be better if that was a part of the dev mode setup instead of direct extension dependencies. Perhaps as a follow up to this PR.

**Comment by aloubyansky** [comment]
I'm wondering whether this could turn into a `dev-console` extension the runtime artifact of which would depend on `vertx-web` and `quarkus-vertx-http` and the deployment one would depend on `quarkus-vertx-http-dev-console-spi`. Then in dev mode the dependency on the `dev-console` extensions would be injected into the app as a direct dependency.

**Comment by aloubyansky** [comment]
E.g. using `QuarkusBootstrap.Builder.setForcedDependencies(...)` in `DevModeMain.start()`.

**Comment by geoand** [comment]
> E.g. using `QuarkusBootstrap.Builder.setForcedDependencies(...)` in `DevModeMain.start()`.

Sounds like an interesting idea!

**Comment by FroMage** [comment]
@aloubyansky you feel like doing this?

**Comment by aloubyansky** [comment]
I'm still on PTO until the 11th. I did look into it a bit before the NY. It looks like the new deps can't be completely removed because they are necessary for compilation. Their scope could change to `provided` to avoid pulling them for the build. Then in dev mode they would be provided by the Quarkus bootstrap. So it's not as elegant as I originally thought. OTOH, it's not entirely unreasonable. That's definitely not a blocker for this PR to be merged.

**Comment by FroMage** [comment]
Test failure is:

```
Error:  Tests run: 6, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.897 s <<< FAILURE! - in io.quarkus.resteasy.reactive.jsonb.deployment.test.sse.SseTestCase
Error:  io.quarkus.resteasy.reactive.jsonb.deployment.test.sse.SseTestCase.testJsonMultiFromMulti  Time elapsed: 0.073 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <2> but was: <0>
	at org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:55)
	at org.junit.jupiter.api.AssertionUtils.failNotEqual(AssertionUtils.java:62)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:150)
	at org.junit.jupiter.api.AssertEquals.assertEquals(AssertEquals.java:145)
	at org.junit.jupiter.api.Assertions.assertEquals(Assertions.java:510)
	at io.quarkus.resteasy.reactive.jsonb.deployment.test.sse.SseTestCase.testJsonMulti(SseTestCase.java:111)
	at io.quarkus.resteasy.reactive.jsonb.deployment.test.sse.SseTestCase.testJsonMultiFromMulti(SseTestCase.java:103)
```

So, SSE, not related to DEV console.

File: extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CaffeineCacheSupplier.java
Comment (FroMage) [reply extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CaffeineCacheSupplier.java]: OK fixed.

**Comment by FroMage** [comment]
OK thanks @aloubyansky it can indeed wait. I've rebased on master to fix the conflict.

**Comment by FroMage** [comment]
Resolved conflict and rebased. We should merge this given that all tests passed last time.

I do need a review though @geoand :)

**Review by geoand — APPROVED** [review]
I didn't re-review, but I was part of the initial development effort, so I'm obviously +1 here

**Comment by gsmet** [comment]
I force pushed to fix the formatting issue. If CI passes, I will merge it so that it's in CR1 and will review after CR1 is released.

**Comment by FroMage** [comment]
Thanks!!!

**Pull request merged**
